extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,321 @@
+
+    @Override
+    public void setContext(@Nullable final SeekableChannelsProvider.ChannelContext context) {
+        // null context equivalent to clearing the context
+        if (context != null && !(context instanceof S3ChannelContext)) {
+            throw new IllegalArgumentException(
+                    "context must be null or an instance of ChannelContext, provided context" +
+                            " of class " + context.getClass().getName());
+        }
+        this.context = (S3ChannelContext) context;
+    }
+
+    @Override
+    public int read(@NotNull final ByteBuffer destination) throws ClosedChannelException {
+        Assert.neqNull(context, "context");
+        Assert.neq(context, "context", SeekableChannelsProvider.ChannelContext.NULL,
+                "SeekableChannelsProvider.ChannelContext.NULL");
+        if (!destination.hasRemaining()) {
+            return 0;
+        }
+        final long localPosition = position;
+        checkClosed(localPosition);
+
+        // Fetch the file size if this is the first read
+        populateSize();
+        if (localPosition >= size) {
+            // We are finished reading
+            return -1;
+        }
+
+        // Send async read requests for current fragment as well as read ahead fragments, if not already in cache
+        final int currFragmentIndex = fragmentIndexForByteNumber(localPosition);
+        final int numReadAheadFragments = Math.min(readAheadCount, numFragmentsInObject - currFragmentIndex - 1);
+        for (int idx = currFragmentIndex; idx <= currFragmentIndex + numReadAheadFragments; idx++) {
+            final CompletableFuture<ByteBuffer> future = context.getCachedFuture(idx);
+            if (future == null) {
+                context.setFragmentContext(idx, sendAsyncRequest(idx));
+            }
+        }
+
+        // Wait till the current fragment is fetched
+        final CompletableFuture<ByteBuffer> currFragmentFuture = context.getCachedFuture(currFragmentIndex);
+        final ByteBuffer currentFragment;
+        try {
+            currentFragment = currFragmentFuture.get(readTimeout.toNanos(), TimeUnit.NANOSECONDS);
+        } catch (final InterruptedException | ExecutionException | TimeoutException | CancellationException e) {
+            final String operation =
+                    "fetching fragment " + currFragmentIndex + " for file " + key + " in S3 bucket " + bucket;
+            throw handleS3Exception(e, operation);
+        }
+
+        // Copy the bytes from fragment from the offset up to the min of remaining fragment and destination bytes.
+        // Therefore, the number of bytes read by this method can be less than the number of bytes remaining in the
+        // destination buffer.
+        final int fragmentOffset = (int) (localPosition - (currFragmentIndex * fragmentSize));
+        currentFragment.position(fragmentOffset);
+        final int limit = Math.min(currentFragment.remaining(), destination.remaining());
```

Comment:
Suggest this rename, since it's not a "limit":
```suggestion
        final int sizeToCopy = Math.min(currentFragment.remaining(), destination.remaining());
```